pkg/virt-controller/watch/export/export.go

```diff
@@ -0,0 +1,878 @@
+		if err != nil {
+			utilruntime.HandleError(err)
+			return
+		} else if exists {
+			ctrl.vmExportQueue.Add(key)
+		}
+	}
+}
+
+func (ctrl *VMExportController) handleService(obj interface{}) {
+	if unknown, ok := obj.(cache.DeletedFinalStateUnknown); ok && unknown.Obj != nil {
+		obj = unknown.Obj
+	}
+
+	if service, ok := obj.(*corev1.Service); ok {
+		key := ctrl.getOwnerVMexportKey(service)
+		_, exists, err := ctrl.VMExportInformer.GetStore().GetByKey(key)
+		if err != nil {
+			utilruntime.HandleError(err)
+			return
+		} else if exists {
+			ctrl.vmExportQueue.Add(key)
+		}
+	}
+}
+
+func (ctrl *VMExportController) getOwnerVMexportKey(obj metav1.Object) string {
+	ownerRef := metav1.GetControllerOf(obj)
+	var key string
+	if ownerRef != nil {
+		if ownerRef.Kind == exportGVK.Kind && ownerRef.APIVersion == exportGVK.GroupVersion().String() {
+			key = controller.NamespacedKey(obj.GetNamespace(), ownerRef.Name)
+		}
+	}
+	return key
+}
+
+func (ctrl *VMExportController) updateVMExport(vmExport *exportv1.VirtualMachineExport) (time.Duration, error) {
+	log.Log.V(3).Infof("Updating VirtualMachineExport %s/%s", vmExport.Namespace, vmExport.Name)
+	var retry time.Duration
+
+	if vmExport.DeletionTimestamp != nil {
+		return retry, nil
+	}
+
+	service, err := ctrl.getOrCreateExportService(vmExport)
+	if err != nil {
+		return 0, err
+	}
+
+	if ctrl.isSourcePvc(&vmExport.Spec) {
+		return ctrl.handleIsSourcePvc(vmExport, service)
+	}
+	return retry, nil
+}
+
+func (ctrl *VMExportController) handleIsSourcePvc(vmExport *exportv1.VirtualMachineExport, service *corev1.Service) (time.Duration, error) {
+	var retry time.Duration
+	pvcs := make([]*corev1.PersistentVolumeClaim, 0)
+	pvc, exists, err := ctrl.getPvc(vmExport.Namespace, vmExport.Spec.Source.Name)
+	if err != nil {
+		return 0, err
+	} else if exists {
+		pvcs = append(pvcs, pvc)
+	}
+
+	pod, exists, err := ctrl.getExporterPod(vmExport)
+	inUse := false
+	if err != nil {
+		return 0, err
+	} else if !exists {
+		inUse, err = ctrl.isPVCInUse(vmExport, pvc)
+		if err != nil {
+			return retry, err
+		}
+		if !inUse && len(pvcs) > 0 {
+			isPopulated, err := ctrl.isPVCPopulated(pvc)
+			if err != nil {
+				return retry, err
+			} else if isPopulated {
+				pod, err = ctrl.createExporterPod(vmExport, pvcs)
+				if err != nil {
+					return 0, err
+				} else if pod == nil {
+					return retry, nil
+				}
+
+				if err := ctrl.getOrCreateCertSecret(vmExport, pod); err != nil {
+					return 0, err
+				}
+			}
+		}
+	} else {
+		if pod.Status.Phase == corev1.PodSucceeded || pod.Status.Phase == corev1.PodFailed {
+			// The server died or completed, delete the pod.
+			ctrl.Recorder.Eventf(vmExport, corev1.EventTypeWarning, exporterPodFailedOrCompletedEvent, "Exporter pod %s/%s succeeded or failed", pod.Namespace, pod.Name)
+			if err := ctrl.Client.CoreV1().Pods(vmExport.Namespace).Delete(context.Background(), pod.Name, metav1.DeleteOptions{}); err != nil {
```

Comment:
what's the goal here for the vmexporter? should it be constantly trying to keep the pod alive if it is failing?

From what I gather here, if the Pod fails, is deleted, or completes for any reason, the pod is deleted and the controller will create a new pod to replace it.

If this is the case, why not use a Deployment of replica 1? then you wouldn't have to think about things like crash loops and backoff... Or is all of this somehow tied to the Cert rotation. So if a pod goes down (because of cert deadline), a new pod with a new secret is created?